test/test.sh

```diff
@@ -0,0 +1,134 @@
+    echo
+    echo '  run             Run the initrd image and kernel image using qemu.'
+    echo '                  Options:'
+    echo '                      --initrd <path>       Path to the initrd image.'
+    echo '                      --kernel <path>       Path to the kernel image.'
+}
+
+# Subcommand: build.
+build() {
+    local output=./initrd_image.cpio
+    local cargo_args=( --cargo-arg=--target=x86_64-unknown-linux-musl )
+
+    while [[ "$#" -gt 0 ]]; do
+        case "$1" in
+            --output) shift; output="$1" ;;
+            --cargo-arg) shift; cargo_args+=("--cargo-arg=$1") ;;
+            *) echo "Unknown option for build: $1"; exit 1 ;;
+        esac
+        shift
+    done
+
+    if [[ -z "$output" ]]; then
+        mkdir -p .tmp
+        output=./.tmp/initrd_image.cpio
+    fi
+
+    # The xtask build-integration-test prints the path to the built image to
+    # stdout. The bluebox command takes arguments for the executable (-e) after
+    # a colon. Use --test-threads=1 to prevent tests from interacting with
+    # shared kernel state due to the lack of inter-test isolation.
+    cargo xtask build-integration-test "${cargo_args[@]}" \
+        | xargs -I {} bluebox -e {}:'--test-threads=1' -o "$output"
+    echo "Built initrd image to $output"
+}
+
+# Subcommand: download-kernel
+download_kernel() {
+    local version=''
+    local output=''
+
+    while [[ "$#" -gt 0 ]]; do
+        case "$1" in
+            --version) shift; version="$1" ;;
+            --output) shift; output="$1" ;;
+            *) echo "Unknown option for download-kernel: $1"; exit 1 ;;
+        esac
+        shift
+    done
+
+    if [[ -z "$version" ]]; then
+        echo "You must specify a kernel version with --version."
+        exit 1
+    fi
+
+    if [[ -z "$output" ]]; then
+        mkdir -p .tmp
+        output="./.tmp/linux-$version.bz"
+    fi
+    url="https://github.com/cilium/ci-kernels/raw/master/linux-${version}.bz"
+    curl -L "$url" -o "$output"
+    echo "Downloaded kernel to $output"
+}
+
+# Subcommand: run
+run() {
+    local initrd=""
+    local kernel=""
+
+    log_file=$(mktemp)
+    trap 'rm -rf "$log_file"' EXIT;
+
+    while [[ "$#" -gt 0 ]]; do
+        case "$1" in
+            --initrd) shift; initrd="$1" ;;
+            --kernel) shift; kernel="$1" ;;
+            *) echo "Unknown option for run: $1"; exit 1 ;;
+        esac
+        shift
+    done
+
+    if [[ -z "$initrd" || -z "$kernel" ]]; then
+        echo "Both --initrd and --kernel options are required for the run command."
+        exit 1
+    fi
+
+    qemu-system-x86_64 \
+        -no-reboot \
+        -append 'printk.devkmsg=on kernel.panic=-1 crashkernel=256M noapic' \
+        -kernel "$kernel" \
+        -initrd "$initrd" \
+        -nographic \
+        -append 'console=ttyS0' \
+        -m '512M' \
+    | tee -a "$log_file"
+    awk '
```

Comment:
This won't do the right thing on non-x86 systems.
I'm also skeptical that it will use hardware assisted virtualization if available - this was all formerly baked into the old script.